### PR TITLE
Add AES implementation in RISC-V 32 Zkn asm

### PR DESCRIPTION
**Pull request by ZenithalHourlyRate**
This PR adds the support of AES implementation for RISC-V 32 platform with Zkn capability. This PR also adds infra for RISC-V 32 platform with CPUID and Configurations (inspired by #17640).

It depends on infra from #17640. It depends on RISCV CAP from #18197. This PR is marked as draft as the infra it depends have not been merged.

The RISC-V 32 infra by this PR is a dependency of the RV32 part of SM4 implementation in #18285.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated
- [x] dependencies (#17640, #18197) are merged

### CODE REVIEW

**Comment by ZenithalHourlyRate** [comment]
Benchmark https://gist.github.com/ZenithalHourlyRate/7b5175734f87acb73d0bbc53391d7140#file-4-openssl32-md

**Comment by marcfedorow** [comment]
Just a small suggestion:
Historically RISC-V ISA assumes 'B' it less than 'K', so zbkb should precede zknd/zkne according to ISA specs, even though it is actually a scalar crypto sub-extension.

**Comment by ZenithalHourlyRate** [comment]
> so zbkb should precede zknd/zkne according to ISA specs

Indeed, I'll fix the function name in the next push. I'll update this PR after #18197 being settled.

File: providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    ossl_ccm_generic_gettag
+};
+
+const PROV_CCM_HW *ossl_prov_aes_hw_ccm(size_t keybits)
+{
+    if (RV32I_ZBKB_ZKND_ZKNE_CAPABLE)
+        return &rv32i_zbkb_zknd_zkne_ccm;
+    else if (RV32I_ZKND_ZKNE_CAPABLE)
+        return &rv32i_zknd_zkne_ccm;
+    else
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc:59]: Drop this `else` and the one two lines up?
They aren't necessary.

File: providers/implementations/ciphers/cipher_aes_gcm_hw_rv32i_zknd_zkne.inc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ossl_gcm_one_shot
+};
+
+const PROV_GCM_HW *ossl_prov_aes_hw_gcm(size_t keybits)
+{
+    if (RV32I_ZBKB_ZKND_ZKNE_CAPABLE)
+        return &rv32i_zbkb_zknd_zkne_gcm;
+    else if (RV32I_ZKND_ZKNE_CAPABLE)
+        return &rv32i_zknd_zkne_gcm;
+    else
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_gcm_hw_rv32i_zknd_zkne.inc:60]: Drop this `else` and the one two lines up.

File: providers/implementations/ciphers/cipher_aes_xts_hw.c [hunk 3]
```diff
@@ -183,6 +183,48 @@ static const PROV_CIPHER_HW aes_xts_rv64i_zknd_zkne = {                        \
...
+};                                                                             \
+static const PROV_CIPHER_HW aes_xts_rv32i_zbkb_zknd_zkne = {                   \
+    cipher_hw_aes_xts_rv32i_zbkb_zknd_zkne_initkey,                            \
+    NULL,                                                                      \
+    cipher_hw_aes_xts_copyctx                                                  \
+};
+# define PROV_CIPHER_HW_select_xts()                                           \
+if (RV32I_ZBKB_ZKND_ZKNE_CAPABLE)                                              \
+    return &aes_xts_rv32i_zbkb_zknd_zkne;                                      \
+else if (RV32I_ZKND_ZKNE_CAPABLE)                                              \
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_xts_hw.c:226]: I don't think this `else` is required.

File: providers/implementations/ciphers/cipher_aes_gcm_hw_rv32i_zknd_zkne.inc [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * RISC-V 32 ZKND ZKNE support for AES GCM.
+ * This file is included by cipher_aes_gcm_hw.c
+ */
+
+static int rv32i_zknd_zkne_gcm_initkey(PROV_GCM_CTX *ctx, const unsigned char *key,
+                                       size_t keylen)
+{
+    PROV_AES_GCM_CTX *actx = (PROV_AES_GCM_CTX *)ctx;
+    AES_KEY *ks = &actx->ks.ks;
+    GCM_HW_SET_KEY_CTR_FN(ks, rv32i_zkne_set_encrypt_key, rv32i_zkne_encrypt,
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_gcm_hw_rv32i_zknd_zkne.inc:20]: Blank line above to separate declarations and code?  Plus more of the same elsewhere.

**Review by paulidale — COMMENTED** [review]
Little nits mostly.  Could somehow who understands RISC-V assembly check aes-riscv32-zkn.pl please?

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
I suppose by the `[WILL BE REBASED]` you've meant this should be squashed into previous commit. Instead of that you can do `git commit --fixup <commit-id>` which will help with automatically squashing fixup commits on rebase when merging.

**Comment by ZenithalHourlyRate** [comment]
> I suppose by the `[WILL BE REBASED]` you've meant this should be squashed into previous commit. Instead of that you can do `git commit --fixup <commit-id>` which will help with automatically squashing fixup commits on rebase when merging.

Yes, I think that's for ease of reviewing. Should I always force-push when making changes or I do small commits and fixup them right before the final merge?

**Comment by t8m** [comment]
> Yes, I think that's for ease of reviewing. Should I always force-push when making changes or I do small commits and fixup them right before the final merge?

The best is to use the git commit --fixup or git commit --squash for the fixup commits. You do not need to do any squashing before the final merge yourself we will do it when merging.

**Comment by ZenithalHourlyRate** [comment]
> Tests will be done shortly. There are two code paths (with `zbkb` and without `zbkb`) so both of them should be tested.

I am quite busy recently. I'm thinking of adding RISC-V 32 CI along side the test of this PR but did not get the time to implement it.

> This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

I assume this PR needs tests, before it requires action by openssl.

File: crypto/aes/asm/aes-riscv32-zkn.pl [hunk 5]
```diff
@@ -0,0 +1,1061 @@
...
+my ($INP,$OUTP,$KEYP) = use_regs(10..12);
+
+# Registers initially to hold Key
+my ($T0,$T1,$T2,$T3) = use_regs(13..16);
+
+# Loop counter
+my ($loopcntr) = use_regs(30);
+
+################################################################################
+# Utility for rv32i_zksed_encrypt and rv32i_zksed_decrypt
```
Comment (marcfedorow) [inline crypto/aes/asm/aes-riscv32-zkn.pl:171]: It looks strange that there are zksed stuff in AES, which is zkn (n is for NIST) but not zks (s is for ShangMi)

**Review by marcfedorow — CHANGES_REQUESTED** [review]

File: crypto/aes/asm/aes-riscv32-zkn.pl (hunk 5 above)
Comment (ZenithalHourlyRate) [reply crypto/aes/asm/aes-riscv32-zkn.pl:171]: That was accidentally copied from https://github.com/openssl/openssl/pull/18285. I'll fix it soon.

**Comment by marcfedorow** [comment]
Side note: SHA2 has a compile-time switch between zknh and no-zknh.
Do you have an intention of rewriting it to run-time logic (as you did it in current PR for AES)?

**Comment by ZenithalHourlyRate** [comment]
> Side note: SHA2 has a compile-time switch between zknh and no-zknh. Do you have an intention of rewriting it to run-time logic (as you did it in current PR for AES)?

See `Discussion on runtime selection` section of #18287

**Comment by marcfedorow** [comment]
This patch works only if both _zkne and zknd are in OPENSSL_riscvcap=rv32imafcp_zkn.
Let's say we do not care about zknd/zkne-only implemeltations (not sure if it is good actually but current structure requires both of them).
The probrem is: there are two different ways to say that we have both of them: `zk` (stands for the whole scalar crypto extension) and `zkn` (stands for all NIST instructions).

(The following text is just an example; linux-only).
I do not want to set this variable manually (i.e. by some external knoweledge of capabilities).
Thus I think reasonable to get it somehow from /proc/cpuinfo.
Let us say that I know that every hart is similar (i.e. has or has not zknd and zkne) or it is a single-core processor.
I may run this:
```bash
read OPENSSL_riscvcap <<<
    $(cat /proc/cpuinfo | grep isa | head -n 1 | awk '{print $3}')
&& export OPENSSL_riscvcap
```
Now I have all the capabilities in my variable.
This is an actual content of /proc/cpuinfo of some proc:
```bash
$ cat /proc/cpuinfo
processor       : 0
hart            : 0
isa             : rv32imafc_zba_zbb_zbc_zbs_zkn_zks_zkt
<...>
```
I.e. it supports bitmanip and all the kcrypto insns (incl. zknd and zkne) but does not support zkr (entropy register).
This string does not have zknd and zkne, so OpenSSL thinks that zkn-based solution can not be used.

If this may be somehow fixed in this PR, it could be great.

**Comment by marcfedorow** [comment]
> See Discussion on runtime selection section of https://github.com/openssl/openssl/pull/18287
Got it, thanks.

**Comment by ZenithalHourlyRate** [comment]
> Thus I think reasonable to get it somehow from /proc/cpuinfo.

Well, the problem is that the current mainline Linux kernel just does not emit `zk`, `zkn` or even `zknd`/`zkne` in cpuinfo. See https://github.com/torvalds/linux/blob/353f7988dd8413c47718f7ca79c030b6fb62cfe5/arch/riscv/kernel/cpu.c#L89-L112. It is too early to add `zk`/`zkn` parsing in OpenSSL side.

I have tried to add at least `zkne` and `zknd` in [a patch](https://lore.kernel.org/linux-riscv/YqY0aSngjI0Hc5d4@Sun/), which does not emit `zk` or `zkn` for now for the simplicity of review. It has not been accepted, though.

>There currently lacks a mechanism to merge them when
>producing cpuinfo. Namely if you provide a riscv,isa
>"rv64imafdc_zk_zks", the cpuinfo output would be
>"rv64imafdc_zbkb_zbkc_zbkx_zknd_zkne_zknh_zkr_zksed
>_zksh_zkt"

After this patch being accepted (or similar patches) and eventually `zk`/`zkn` being able to be emitted from cpuinfo, I think then it will be the time to add `zk`/`zkn` parsing in OpenSSL, as the env var mechanism now needs human intervention so providing either `zknd_zkne` or `zk_zkn` needs the same effort for end users. Or, OpenSSL could use the HWCAP method proposed in that patch so there is no need to parse `zk`/`zkn`.

**Comment by marcfedorow** [comment]
Thank you very much for clarifications!

**Comment by marcfedorow** [comment]
Another problem with riscv-isa-string parsing.
Currently there are zero legal isa strings that causes an actual problem with OPENSSL_riscvcap logic.
But that's more a coincidence than a robust parsing.
This is so because of:
(1) OpenSSL requires at least two subextensions;
(2) OpenSSL checks the lowest subextensions (i.e. zkne, not zkn).

There are two problems:
(1) Parser does not understand if the first 'z' is a part of a name (i.e. rv32ixzkne sets riscvcap[zkne] to 1);
(2) rv32i_zbkbq_zkndq_zkneq sets riscvcap[zbkb, zknd, zkne] to 1 while neither of them is presented in the isa string.

While it works by luck in current circumstances, it should be nevertheless fixed in my opinion.

File: providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+/*-
+ * RISC-V 32 ZKND ZKNE support for AES CCM.
+ * This file is included by cipher_aes_ccm_hw.c
+ */
+
+static int ccm_rv32i_zknd_zkne_initkey(PROV_CCM_CTX *ctx, const unsigned char *key,
+                             size_t keylen)
+{
+    PROV_AES_CCM_CTX *actx = (PROV_AES_CCM_CTX *)ctx;
+
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc:19]: Spurious blank line.

File: providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                          NULL, NULL);
+    return 1;
+}
+
+static int ccm_rv32i_zbkb_zknd_zkne_initkey(PROV_CCM_CTX *ctx, const unsigned char *key,
+                             size_t keylen)
+{
+    PROV_AES_CCM_CTX *actx = (PROV_AES_CCM_CTX *)ctx;
+
+    AES_HW_CCM_SET_KEY_FN(rv32i_zbkb_zkne_set_encrypt_key, rv32i_zkne_encrypt,
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc:30]: Spurious blank line.

File: providers/implementations/ciphers/cipher_aes_hw_rv32i_zknd_zkne.inc [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+#define cipher_hw_rv32i_zbkb_zknd_zkne_cbc    ossl_cipher_hw_generic_cbc
+#define cipher_hw_rv32i_zbkb_zknd_zkne_ecb    ossl_cipher_hw_generic_ecb
+#define cipher_hw_rv32i_zbkb_zknd_zkne_ofb128 ossl_cipher_hw_generic_ofb128
+#define cipher_hw_rv32i_zbkb_zknd_zkne_cfb128 ossl_cipher_hw_generic_cfb128
+#define cipher_hw_rv32i_zbkb_zknd_zkne_cfb8   ossl_cipher_hw_generic_cfb8
+#define cipher_hw_rv32i_zbkb_zknd_zkne_cfb1   ossl_cipher_hw_generic_cfb1
+#define cipher_hw_rv32i_zbkb_zknd_zkne_ctr    ossl_cipher_hw_generic_ctr
+
+static int cipher_hw_rv32i_zknd_zkne_initkey(PROV_CIPHER_CTX *dat,
+                                   const unsigned char *key, size_t keylen)
```
Comment (paulidale) [inline providers/implementations/ciphers/cipher_aes_hw_rv32i_zknd_zkne.inc:32]: The indentation here looks off by one.

File: crypto/aes/asm/aes-riscv32-zkn.pl (hunk 5 above)
Comment (paulidale) [reply crypto/aes/asm/aes-riscv32-zkn.pl:171]: ?

**Review by paulidale — APPROVED** [review]

File: crypto/aes/asm/aes-riscv32-zkn.pl (hunk 5 above)
Comment (ZenithalHourlyRate) [reply crypto/aes/asm/aes-riscv32-zkn.pl:171]: > I am quite busy recently.

Sorry about this. I will finish my deadline in 3-5 days so I'll fix it and other commends by the end of this month.

**Comment by ZenithalHourlyRate** [comment]
Tests passed. I used the master branch of QEMU and set up `qemu-riscv32` in binfmt, then compiled this branch with the latest riscv-gnu-toolchain (riscv32-glibc-ubuntu-20.04-nightly-2022.08.26-nightly.tar.gz)

```bash
# compile
CC="riscv32-unknown-linux-gnu-gcc" ./config --banner=Configured linux32-riscv32
make -j12
# test
export QEMU_CPU="rv32,zbb=true,zbkb=true,zknd=true,zkne=true"
export OPENSSL_riscvcap="rv32gc_zbkb_zknd_zkne"
make test -j12 HARNESS_JOBS=12 TESTS="-test_afalg" QEMU_LD_PREFIX=/opt/riscv/sysroot
# result
All tests successful.
Files=252, Tests=3425, 779 wallclock secs (11.00 usr  1.12 sys + 5051.56 cusr 78.72 csys = 5142.40 CPU)
Result: PASS
```

I also covered another code branch, i.e. with zknd/zkne but without zbkb
```
# test
export QEMU_CPU="rv32,zbb=false,zbkb=false,zknd=true,zkne=true" 
export OPENSSL_riscvcap="rv32gc_zknd_zkne"
make test -j12 HARNESS_JOBS=12 TESTS="-test_afalg" QEMU_LD_PREFIX=/opt/riscv/sysroot
```

A special note on `enable-fips` config is that it would fail the first test which does `fips install`, saying that `QEMU_CPU` and `OPENSSL_riscvcap` are useless env vars.

Also, I've tried to add riscv32 CI but it seems that ubuntu-latest does not have `libc6-dev-riscv32-cross` and corresponding gcc so the environment could not be conveniently set up (or should we download the toolchain/rootfs from [riscv-collab/riscv-gnu-toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain/)?). Also the QEMU in ubuntu-latest is too old to support Zk extension.

As for the issues mentioned in https://github.com/openssl/openssl/pull/18308#issuecomment-1192482508, I will address them in another PR.

File: providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc (hunk 6 above)
Comment (ZenithalHourlyRate) [reply providers/implementations/ciphers/cipher_aes_ccm_hw_rv32i_zknd_zkne.inc:19]: https://github.com/openssl/openssl/blob/56233ba8574c01b3912cf662335fedaabc7faec2/providers/implementations/ciphers/cipher_aes_ccm_hw_aesni.inc#L18-L20

This is copy-pasted from another file.

File: providers/implementations/ciphers/cipher_aes_hw_rv32i_zknd_zkne.inc (hunk 8 above)
Comment (ZenithalHourlyRate) [reply providers/implementations/ciphers/cipher_aes_hw_rv32i_zknd_zkne.inc]: Fixed now.

File: crypto/aes/asm/aes-riscv32-zkn.pl (hunk 5 above)
Comment (ZenithalHourlyRate) [reply crypto/aes/asm/aes-riscv32-zkn.pl]: Fixed now.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]
Just a formal approval - the actual assembler code not reviewed.

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**